File: LeetCode/1937. Maximum Number of Points with Cost/solution.cpp
const int inf = 0x3f3f3f3f;
const int ninf = -inf;

class Solution {
public:
    long long maxPoints(vector<vector<int>>& points) {
        vector<long long> dp(points[0].begin(), points[0].end());
        int n = points.size(), m = points[0].size();
        for (int i = 1; i < n; i++){
            vector<long long> ndp(m, ninf);
            long long mx = ninf;
            for (int j = 0; j < m; j++){
                mx = max(mx, dp[j]+j);
                ndp[j] = max(ndp[j], mx-j+points[i][j]);
            }
            mx = ninf;
            for (int j = m-1; j >= 0; j--){
                mx = max(mx, dp[j]-j);
                ndp[j] = max(ndp[j], mx+j+points[i][j]);
            }
            std:swap(dp, ndp);
        }
        return *max_element(dp.begin(), dp.end());
        
    }
};

File: LeetCode/1937. Maximum Number of Points with Cost/solution.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        n, m = len(points), len(points[0])
        dp = points[0][:]
        for i in range(1,n):
            leftmx, right = -1, []
            ndp = []
            for j in range(m):
                while right and dp[right[-1]]-right[-1] <= dp[j]-j:
                    right.pop()
                right.append(j)
            right = right[::-1]
            for j in range(m):
                if leftmx == -1 or dp[leftmx]+leftmx <= dp[j]+j:
                    leftmx = j  
                while right and right[-1] < j:
                    right.pop()
                ndp.append(max(dp[leftmx]+leftmx-j,dp[right[-1]]-right[-1]+j)+points[i][j])
            dp = ndp
            # print(dp)
        return max(dp)
